### ISSUE TITLE
CPU usage in idle mode question

### ISSUE THREAD

**Issue by gavrilyak** [maia-sdr/maia-sdr#29]
maia-httpd uses 30% of CPU when no clients connected (number_integrations = 16) and above 50% when   number_integrations = 1. Pluto is  warm.
I added the following condition and CPU usage drops to 5% on number_integrations = 1.
```rust
  if self.sender.receiver_count() > 0 {
    let _ = self.sender.send(Self::buffer_u64_to_f32(buffer, scale));
  }
``` 

So my question is: is this a proper fix? Maybe it will break something or maybe it could be done better (e.g. don't read buffers at all when no client is connected) .

Thanks in advance.

**Comment by daniestevez** [comment]
What sample rate are you using? 16 or 1 integrations is a rather small value for any reasonable sample rate. It is quite possible that you'll lose spectrometer output buffers at this rate.

Indeed there are a number of optimizations that could be done when no client is connected. Your suggestion avoids a conversion from u64 to f32, but we could even ignore the spectrometer completely (as you also mention), though we will still be waken up by its interrupts. Another possible improvement would be to completely disable the spectrometer when no clients are connected, though the HDL doesn't currently have an option for this.

**Comment by gavrilyak** [comment]
I am using 56 MSpS, and you are absolutely right, I am loosing spectrometer data in browser.  So yeah, it's just a possible enhancement. 
Actually I am not trying to use those spectrometer data in browser, I want to use them in Pluto itself to detect signals and measure their duration, that's why I am trying so crazy small integrations. For me it would be better to have smaller FFT width (e.g 128 or even 64), but it seems  that I need to build FPGA binary for this.  I tried to do it, but no luck with downloading Vivado, because of export control and Ukraine. So for now I am "integrating" spectrum data on CPU instead of integrating signals on FPGA, as I should. 

Thank you for the awesome project, it really opens new possibilities for SDR development. Have a very good luck!

**Comment by daniestevez** [comment]
Now I understand why you care about the case when no clients are connected to maia-httpd.

Reading the documentation for [Tokio broadcast channels](https://docs.rs/tokio/latest/tokio/sync/broadcast/index.html), I think your suggested modification is alright. The "worst" thing that would happen that the number of receivers changes between the moment when we check it and the moment when we call `send`. But this won't cause any problems.

A pull request with this change would be welcome.

**Comment by gavrilyak** [comment]
Here is a pull request with those exact changes.
I also tried to  ignore the spectrometer completely, by doing this check for receivers count right after interrupt.wait(), but this doesn't affect CPU usage much. Also reading from buffers have a side effect of cache invalidation, and I do not quite understand how it works, so I decided it's better to check for receivers exactly when we are going to send packet (better code locality or something...).
